### ISSUE TITLE
NUT SSL Certificate support

### ISSUE THREAD

**Issue by Geraden07** [collectd/collectd#1760]
I'm running collectd Version 5.5.1 Release 2.fc23 (Fedora 23 current package).

I have successfully configured the NUT plugin to monitor the upsd server for my UPS. Everything works.

However, in the journal when collectd starts up, I see the following:

```
<snip>
Connected to localhost in SSL
nut plugin: Connection to (localhost, 3493) established.
Certificate verification is disabled
<snip>
```

I have verified that this message disappears when I disable only the NUT Plugin. But the man page doesn't list any other options for configuring the nut plugin besides the "UPS" directive. I've tried "VerifyPeer" and "VerifyHost" directives and neither work, both cause errors in the log.

Is this functionality that is simply not implemented yet? Is it coming?

**Comment by rubenk** [comment]
Hi,

It seems you're right. Looking at the code this functionality is indeed not implemented yet.
I'm not too familiar with nut, but I checked the api documentation and the nut library doesn't seem to have support for specifying certificates.

http://networkupstools.org/docs/user-manual.chunked/ar01s09.html suggests that you can specify certificates in your upsmon.conf, might be worth a try.

**Comment by Geraden07** [comment]
Nut has two components: The nut-server (aka upsd) and nut-monitor (aka upsmon).
(In reality there is a third, the driver that communicates with the device, but that is managed by the nut-server so not of concern).

upsd makes the information from the device available to anything that connects to it. The upsmon connects and monitors changes in the device info (ie a battery critical event) and takes actions upon those changes.

It's my understanding that collectd does not interface at all with upsmon, but instead polls and collects information from the upsd server directly.

My upsd server is configured to serve over SSL with cert and private key. My upsmon functions with this.

My collectd accepts the server's cert and can talk successfully with the upsd server, but the warning it shows (that I mentioned in my first post) indicates that it blindly accepts this certificate without doing a cert check against a predefined CA.

My impression is that for this to be implemented, the collectd-nut plugin would need to at least add the ability to define a CA that could be used to verfiy the certificate it receives from the upsd server.

My understanding of how this is done in other plugins (for example the collectd-apache plugin, see the man here: https://collectd.org/documentation/manpages/collectd.conf.5.shtml#plugin_apache ) is that the plugin config can specify 4 directives of note:

```VerifyPeer  true/false``` tells collectd whether to verify if a cert is signed by the CA.
```VerifyHost  true/false``` tells collectd whether to verify that the host name it is talking to and the host name in the certificate it received match.
```CACert <path>``` tells collectd where to look for the CA cert to use for authenticating a certificate.
```SSLCiphers <list of ciphers>``` this one would be optional. It allows a list of ciphers to be explicitly defined. This gives the ability to disallow insecure and old ciphers by not including them in the list, possibly preventing downgrade attacks for example.

I imagine something similar would be ideal for collectd-nut.

**Comment by rubenk** [comment]
Thanks for the explanation.

Collectd uses libupsclient for communicating with upsd. There seems to be some support for setting certificates in that library: http://networkupstools.org/docs/man/upscli_add_host_cert.html, so this should be possible.

**Comment by Geraden07** [comment]
(See last section for TL;DR)
So I've taken a stab at implementing this myself. I broke it down into three steps:

1. Implement ForceSSL flag.
2. Implement VerifyPeer (aka CAVerify) flag, along with CAPath.
3. Implement VerifyHost flag.

Once I dove into the code I was able to trace what was happening where... Obviously the collectd code (in nut.c) calls functions from the libupsclient library. That library provides the following: 
* `upscli_connect` (man page: http://networkupstools.org/docs/man/upscli_connect.html )
* `upscli_add_host_cert` (man page: http://networkupstools.org/docs/man/upscli_add_host_cert.html )
* `upscli_init` (man page: http://networkupstools.org/docs/man/upscli_init.html )

From there, the libupsclient will call EITHER OpenSSL or NSS depending on which it was compiled with. On Fedora25, the version provided by the package manager was compiled with OpenSSL.

In the OpenSSL version, the `upscli_add_host_cert` does nothing. It is literally an empty definition. Which is a shame because that method WOULD provide parameters for everything we want (hostname, certname, verifycert, requiressl). Oh well.

## 1. ForceSSL

This went fairly smoothly. We already call `upscli_connect` in the current code, and it takes a flag argument UPSCLI_CONN_TRYSSL which will attempt to connect with SSL and fallback to unencrypted if it cannot. The other flag documented is UPSCLI_CONN_REQSSL which requires SSL and won't let it fall back, erroring out instead.

So all I had to do was implement a new config value, and use it to decide which flag to pass. I tested and it works perfectly. On to the next bit...

## 2. VerifyPeer / CAVerify with CAPath

I looked at the `upscli_init` call. It provides parameters for certverify and certpath. It also provides two extra parameters certname and certpasswd but these are unused if libupsclient is compiled with OpenSSL, so we can ignore them.

There was a lot of trial and error, and me code diving into not just libupsclient but also openssl (a very little bit). I discovered a couple important things:
* Contrary to the libupsclient documentation, the capath variable we must provide is for a **folder** not a certificate file.
* The capath folder must contain certificates/links named as XXXXXXXX.Y where X is their hash and Y is an arbitrary int (usually 0) that can be incremented to avoid collisions of two different certs that may have the same hash.
* There is an additional undocumented flag we can pass `upscli_connect` called UPSCLI_CONN_CERTVERIF which requires the certificate be verified. This can be combined with the REQSSL flag using bitwise OR.

A handy utility for creating hash-named links to all certificates in a directory is c_rehash (provided by openssl-perl). Otherwise the command to do it manually is:
```ln -s <CERTFILE> ./$(openssl x509 -hash -noout -in <CERTFILE>).0```

Once all that was figured out and implemented, I was able to do exhaustive tests (see attached Excel spreadsheet in TL;DR section) for every combination of config inputs. I was able to toggle the config of my upsd server so I could test against against a server with and without SSL enabled for each scenario to ensure proper behaviour.

## 3. VerifyHost

Lastly... Unfortunately this is the only feature that doesn't seem to be possible right now with libupsclient. Unless we were to somehow specify that the version of the library required was the one compiled with NSS, we can't be sure of the `upscli_add_host_cert` call actually doing anything.

The OTHER way we could implement this would be to `#include <openssl/ssl.h>` in nut.c and then we could call `X509 cert = SSL_get_peer_certificate(ups->conn->ssl);` to get the certificate received. We could then string compare the cert's hostname against the `ups->hostname` to verify that they match.

I dislike this approach though since it adds the dependency to OpenSSL directly from nut.c . If ever the libupsclient were to change underlying libraries, or the version installed was compiled with NSS instead of OpenSSL, I'm unsure what would happen then with nut.c attempting to use OpenSSL to get the certificate from a connection that OpenSSL did not create.

I think choosing not to implement this particular flag is safest at this point. It's not a big thing to give up, since we are still able to accomplish the verification and forcing SSL.

# TL;DR
I've successfully implemented ForceSSL, VerifyPeer and CAPath, but not VerifyHost, which seems to be impossible right now without introducing additional dependency and compatibility issues.

I've tested extensively and documented all possible combinations of inputs for these new config values against both a upsd server with SSL available, and without. See attached Excel spreadsheet for test scenario breakdown.
[collectd-nut Unit Testing.xlsx](https://github.com/collectd/collectd/files/721872/collectd-nut.Unit.Testing.xlsx)

I've read and complied with (I think) all the recommendations in the CONTRIBUTING.md file in the project root. I'll open a pull request with these changes and reference this issue. If anyone needs a upsd server to test this against, I would be open to coordinating with them and opening a port on my network for them to use my server. UPS's aren't cheap and if someone doesn't have one, I'm not sure how easy it is to get a upsd server up and running with no device to connect to.

**Comment by rubenk** [comment]
Fixed by 4138e6c7622acf45d8e441a2d5085190b832c439
